pkg/controller/volume/attachdetach/populator/desired_state_of_world_populator.go

```diff
@@ -113,3 +122,208 @@ func (dswp *desiredStateOfWorldPopulator) findAndRemoveDeletedPods() {
+}
+
+// notRunning returns true if every status is terminated or waiting, or the status list
+// is empty.
+func notRunning(statuses []v1.ContainerStatus) bool {
+	for _, status := range statuses {
+		if status.State.Terminated == nil && status.State.Waiting == nil {
+			return false
+		}
+	}
+	return true
+}
+
+func (dswp *desiredStateOfWorldPopulator) isPodVolumePreviouslyAdded(podName volumetypes.UniquePodName, volumeSpec *volume.Spec, nodeName types.NodeName) (bool, error) {
+	return dswp.desiredStateOfWorld.IsPodVolumePreviouslyAdded(podName, volumeSpec, nodeName)
+}
+
+func (dswp *desiredStateOfWorldPopulator) findAndAddActivePods() {
+	pods, err := dswp.podLister.List(labels.Everything())
+	if err != nil {
+		glog.Errorf("podLister List failed for pods %v", err)
+		return
+	}
+	for _, pod := range pods {
+		if dswp.isPodTerminated(pod) {
+			// Do not add volumes for terminated pods
+			continue
+		}
+
+		nodeName := types.NodeName(pod.Spec.NodeName)
+		if nodeName == "" {
+			glog.V(10).Infof(
+				"Skipping processing of pod %q/%q: it is not scheduled to a node.",
+				pod.Namespace,
+				pod.Name)
+			continue
+		} else if !dswp.desiredStateOfWorld.NodeExists(nodeName) {
+			// If the node the pod is scheduled to does not exist in the desired
+			// state of the world data structure, that indicates the node is not
+			// yet managed by the controller. Therefore, ignore the pod.
+			glog.V(10).Infof(
+				"Skipping processing of pod %q/%q: it is scheduled to node %q which is not managed by the controller.",
+				pod.Namespace,
+				pod.Name,
+				nodeName)
+			continue
+		}
+		// check if the pod has volumes
+		if len(pod.Spec.Volumes) <= 0 {
+			glog.V(10).Infof("Skipping processing of pod %q/%q: it has no volumes.",
+				pod.Namespace,
+				pod.Name)
+		} else {
+			for _, podVolume := range pod.Spec.Volumes {
+				uniquePodName := volumehelper.GetUniquePodName(pod)
+				volumeSpec, err := dswp.createVolumeSpec(podVolume, pod.Namespace)
+				if err != nil {
+					glog.V(10).Infof(
+						"Error processing volume %q for pod %q/%q: %v",
+						podVolume.Name,
+						pod.Namespace,
+						pod.Name,
+						err)
+					continue
+				}
+
+				// check if the pod and its volume is added to dsw
+				added, err := dswp.isPodVolumePreviouslyAdded(uniquePodName, volumeSpec, nodeName)
+				if err != nil {
+					glog.V(10).Infof("Check whether pod references the specified volume is added to dsw falied: %v", err)
+					continue
+				} else if added {
+					glog.V(10).Infof("Pod references the specified volume is already added to dsw")
+					continue
+				}
+
+				dswp.desiredStateOfWorld.AddPod(uniquePodName, pod, volumeSpec, nodeName)
+			}
+		}
+	}
+
+}
+
+// createVolumeSpec creates and returns a volume.Spec object for the
+// specified volume. It dereference any PVC to get PV objects, if needed.
+func (dswp *desiredStateOfWorldPopulator) createVolumeSpec(podVolume v1.Volume, podNamespace string) (*volume.Spec, error) {
```

Comment:
This logic is duplicated here and in `pkg/controller/volume/attachdetach/attach_detach_controller.go`. Instead of duplicating the code, let's move the shared logic into a separate package (maybe `pkg/controller/volume/attachdetach/util/`?) that both can import.